go/vt/vtgate/schema/tracker.go

```diff
@@ -65,39 +67,86 @@ const defaultConsumeDelay = 1 * time.Second
 	// Set the caller on the context if the user is provided.
 	// This user that will be sent down to vttablet calls.
 	if user != "" {
 		ctx = callerid.NewContext(ctx, nil, callerid.NewImmediateCallerID(user))
 	}
 
-	return &Tracker{
+	t := &Tracker{
 		ctx:          ctx,
 		ch:           ch,
 		tables:       &tableMap{m: map[keyspaceStr]map[tableNameStr][]vindexes.Column{}},
 		tracked:      map[keyspaceStr]*updateController{},
 		consumeDelay: defaultConsumeDelay,
 	}
+
+	if enableViews {
+		t.views = &viewMap{m: map[keyspaceStr]map[viewNameStr]sqlparser.SelectStatement{}}
+	}
+	return t
 }
 
 // LoadKeyspace loads the keyspace schema.
 func (t *Tracker) LoadKeyspace(conn queryservice.QueryService, target *querypb.Target) error {
-	res, err := conn.Execute(t.ctx, target, mysql.FetchTables, nil, 0, 0, nil)
+	err := t.loadTables(conn, target)
+	if err != nil {
+		return err
+	}
+	err = t.loadViews(conn, target)
+	if err != nil {
+		return err
+	}
+
+	t.tracked[target.Keyspace].setLoaded(true)
+	return nil
+}
+
+func (t *Tracker) loadTables(conn queryservice.QueryService, target *querypb.Target) error {
+	if t.tables == nil {
+		// this can only happen in testing
+		return nil
+	}
+
+	ftRes, err := conn.Execute(t.ctx, target, mysql.FetchTables, nil, 0, 0, nil)
 	if err != nil {
 		return err
 	}
 	t.mu.Lock()
 	defer t.mu.Unlock()
+
 	// We must clear out any previous schema before loading it here as this is called
 	// whenever a shard's primary tablet starts and sends the initial signal. Without
 	// clearing out the previous schema we can end up with duplicate entries when the
 	// tablet is simply restarted or potentially when we elect a new primary.
 	t.clearKeyspaceTables(target.Keyspace)
-	t.updateTables(target.Keyspace, res)
-	t.tracked[target.Keyspace].setLoaded(true)
-	log.Infof("finished loading schema for keyspace %s. Found %d columns in total across the tables", target.Keyspace, len(res.Rows))
+	t.updateTables(target.Keyspace, ftRes)
+	log.Infof("finished loading schema for keyspace %s. Found %d columns in total across the tables", target.Keyspace, len(ftRes.Rows))
+
+	return nil
+}
+
+func (t *Tracker) loadViews(conn queryservice.QueryService, target *querypb.Target) error {
+	if t.views == nil {
+		// This happens only when views are not enabled.
+		return nil
+	}
+
+	fvRes, err := conn.Execute(t.ctx, target, mysql.FetchViews, nil, 0, 0, nil)
+	if err != nil {
+		return err
+	}
+
+	// We must clear out any previous view definition before loading it here as this is called
+	// whenever a shard's primary tablet starts and sends the initial signal. Without
+	// clearing out the previous view definition we can end up with duplicate entries when the
+	// tablet is simply restarted or potentially when we elect a new primary.
+	t.clearKeyspaceViews(target.Keyspace)
+	t.updateViews(target.Keyspace, fvRes)
+	log.Infof("finished loading views for keyspace %s. Found %d views", target.Keyspace, len(fvRes.Rows))
+
 	return nil
 }
```

Comment:
`loadTables` takes a lock on `t.mu`, shouldn't `loadViews` also take the same lock?